Move predict out of config if possible

ways opened the pull request:
Should be merged after https://github.com/metno/bris-inference/pull/28/files

havardhhaugen commented on bris/schema/schema.json:
```diff
@@ -126,14 +126,41 @@
+            "type": "string"
+        },
+        "dataset": {
+            "description": "Input dataset. Can be a single path, or two datasets merged by cutout. Example dataset.cutout = ['/path/dataset1', '/path/dataset2' ]",
```
Make this part less strict, since we have to support dataset types beyond single dataset and cutout. This field should contain a valid anemoi datasets recipe, and then we probably have to leave it up to the user to make sure this matches what they used to train the model. In principle the recipe in training can differ from the recipe during inference if a different dataset is used for inference.

ways replied:
Fixed in cd44df5..b31b00e

ways commented:
This runs for me with config

```
start_date: 2024-12-03T00:00:00
end_date: 2024-12-03T06:00:00

checkpoint_path: /home/larsfp/nobackup/bris/checkpoint/aifs_inference_o96_10k_stretched_grid.ckpt

leadtimes: 2
frequency: 6h

deterministic: True

# If the user wants to release GPU cache and memory
# This option releases unused cached/memory used by torch
release_cache: True

# Determine how much the encoder and decoder is chunked
inference_num_chunks: 32

dataset:
  cutout:
    - dataset: /home/larsfp/nobackup/bris/input/meps_2024120306.zarr
    - dataset: /home/larsfp/nobackup/bris/input/hres_2024120306.zarr
  min_distance_km: 0
  # adjust: all
workdir: /home/larsfp/nobackup/bris/tmp

dataloader:
  # batch_size: 4
  prefetch_factor: 2
  num_workers: 1
  pin_memory: True

  read_group_size: 1 #Do not change this, not implemented properly

  datamodule:
    _target_: anemoi.training.data.dataset.NativeGridDataset
    _convert_: all

hardware:
  num_gpus_per_node: 1
  num_gpus_per_model: 1
  num_nodes: 1

model:
  _target_: bris.model.BrisPredictor
  _convert_: all

routing:
  - decoder_index: 0
    domain_index: 0
    outputs:
      - netcdf:
          filename_pattern: meps_pred_%Y%m%dT%HZ.nc
          variables: [2t, 2d]
  - decoder_index: 0
    domain_index: 1
    outputs:
      - netcdf:
          filename_pattern: era_pred_%Y%m%dT%HZ.nc
          variables: [2t, 2d]
```

ways commented:
Should be ready to go.

Note that I set

> validate(args.config, raise_on_error=True)

which isn't really related to this issue. But I can't understand why one would continue if the (now loosely defined) config validation fails.

havardhhaugen commented:
Deterministic true has nothing to do with this pr, but I might as well correct it now. Not sure why aram set it to False by default, probably a typo. You should be able to remove deterministic: True from the config.

ways commented:
> Deterministic true has nothing to do with this pr, but I might as well correct it now. Not sure why aram set it to False by default, probably a typo. You should be able to remove deterministic: True from the config.

We have this, https://github.com/metno/bris-inference/blob/f36ee364f8c0bbecbc213906b934b0d88f1650d2/bris/inference.py#L35. Do you want that hardcoded to True? Or default to True, will possible override from config?

havardhhaugen commented:
> > Deterministic true has nothing to do with this pr, but I might as well correct it now. Not sure why aram set it to False by default, probably a typo. You should be able to remove deterministic: True from the config.
> 
> The rest is OK and ready for merge?

I have not been able to test it end to end yet due to lumi queue and some config mistakes on my end.

havardhhaugen commented:
> > Deterministic true has nothing to do with this pr, but I might as well correct it now. Not sure why aram set it to False by default, probably a typo. You should be able to remove deterministic: True from the config.
> 
> We have this,
> 
> https://github.com/metno/bris-inference/blob/f36ee364f8c0bbecbc213906b934b0d88f1650d2/bris/inference.py#L35
> 
> . Do you want that hardcoded to True? Or default to True, will possible override from config

Yes I would remove this and hardcode it to True, I just didnt realise it was a property of Inference in my earlier commit but I have removed the self.deterministic from Inference now. From the pytorch lightning documentation:

This flag sets the torch.backends.cudnn.deterministic flag. Might make your system slower, but ensures reproducibility.

Since this package should only be used for inference I dont see a usecase where anyone would want it to be false

havardhhaugen commented:
I'm running into an issue now where it looks like the data is not being passed correctly to the model, will look into it more tomorrow.

ways commented:
> I'm running into an issue now where it looks like the data is not being passed correctly to the model, will look into it more tomorrow.

Can you post logs or output?

ways commented:
Latest commit seems to work from what I can test, and start & end is supplied to open_dataset.

![image](https://github.com/user-attachments/assets/c1993d0d-9c23-4ef7-82ad-0684e19a89e0)

havardhhaugen commented:
I simplified the dataset config setup somewhat, and added frequency. The dict that is passed to open_dataset can be viewed as a hierarchy of operations, and we always want to do the adjust of start and end after whatever goes on to combine the dataset in config.dataset. 
Also added the extension of start_date to account for multi_step, so now the first prediction will be at start_date specified in the config.

havardhhaugen commented:
This runs for me now, so if you are happy with the changes it can be merged

ways commented:
> I simplified the dataset config setup somewhat, and added frequency. The dict that is passed to open_dataset can be viewed as a hierarchy of operations, and we always want to do the adjust of start and end after whatever goes on to combine the dataset in config.dataset. Also added the extension of start_date to account for multi_step, so now the first prediction will be at start_date specified in the config.

Aha. I misunderstood the needed layout. Merging.
